Fix #4143 - add binary and bigint formats (#4159)
MicrosoftDocs/PowerShell-Docs: reference/7/Microsoft.PowerShell.Core/About/about_numeric_literals.md

| l      | long data type                 |                         |
| u      | unsigned int or long data type | Added in PowerShell 6.2 |
| ul     | unsigned long data type        | Added in PowerShell 6.2 |

> | n      | bigint data type               | Added in PowerShell 7.0 |

Reviewer: Given that `BigInteger` is technically a non-primitive type (it's a `byte`-based `struct`, I believe), would it be more appropriate to give its full name here? `System.Numerics.BigInteger`

Author: Updated.

> The following type accelerators were added in PowerShell 6.2: `[short]`,
> `[ushort]`, `[uint]`, `[ulong]`.

> 

Reviewer: Rather than remove this whole section, I might be more inclined to simply mention use of the `n` suffix as a much simpler alternative to the string->cast method. I think there's still value in having it explicitly mentioned like this for users who aren't yet aware that this rounding occurs.

It might also be worth a mention that some values between `[ulong]::MaxValue` and `[decimal]::MaxValue` may also be denoted as `Decimal` by suffix to maintain accuracy. These values are usually parsed as `Double` due to the `real`-parsing mode not considering `decimal` without specifically using the suffix.

Author: Got it.

## Examples

The following table contains several examples of numeric literals and lists
their type and value:

|  Number   |  Type   |    Value     |
| ---------:| ------- | -----------: |

> |       100 | Int32   |          100 |

Reviewer: ```suggestion
|       100 | Int32   |          100 |
|      100u | UInt32  |          100 |
```

Author: done

|      100D | Decimal |          100 |
|      100l | Int64   |          100 |
|     100uL | UInt64  |          100 |
|     100us | UInt16  |          100 |
|     100uy | Byte    |          100 |
|      100y | SByte   |          100 |
|       1e2 | Double  |          100 |
|      1.e2 | Double  |          100 |
|     0x1e2 | Int32   |          482 |
|    0x1e2L | Int64   |          482 |
|    0x1e2D | Int32   |         7725 |
|      482D | Decimal |          482 |
|     482gb | Int64   | 517543559168 |
|  0x1e2lgb | Int64   | 517543559168 |

> | 0b1011011 | Int32   |           91 |

Reviewer: Examples should be added here (and it should be noted somewhere else as well, if we haven't already somewhere below) for how adding the `u` suffix can alter the behaviour of hexadecimal or binary numerals.

Example:

```ps
PS> 0xFFFFFFFF
-1
PS> 0xFFFFFFFFu
4294967295 # [uint]::MaxValue
```

Similar things will happen with binary in 7.

Author: Done.

### Working with binary or hexadecimal numbers

Overly large binary or hexadecimal literals return as `[bigint]` rather than
failing the parse. Sign bits are still respected above even `[decimal]` ranges,
however:

- If a binary string is some multiple of 8 bits long, the highest bit is
  treated as the sign bit.
- If a hex string, which has a length that is a multiple of 8, has the first
  digit with 8 or higher, the numeral is treated as negative.

Sign bits are accepted for bigint-suffixed numerals:

- Bigint-suffixed hex treats the high bit of any literal with a length multiple
  of 8 as the sign bit.
- Bigint-suffixed binary accepts sign bits at 96 and 128 chars, and from there
  on every 8 characters.
- Prefixing the literal with a 0 will bypass this and be treated as unsigned.
  For example: `0b011111111`.

> - Specifying an unsigned suffix ignores sign bits.

Reviewer: ```suggestion
- Specifying an unsigned suffix ignores sign bits. For example, `0xFFFFFFFF` usually returns `-1`, but `0xFFFFFFFFu` will return `[uint]::MaxValue`, the value  `4294967295`.
```

Author: Got it.

- If the string has no type suffix, then parse as `[Double]`.

### Paring integer numeric literals

Integer type literals are parsed using the following steps:

1. Determine the radix format
   - For binary formats, parse into `[BigInteger]`.
   - For hexidecimal formats, parse into `[BigInteger]` using special casies to
     retain original behaviours when the value is in the `[int]` or `[long]`
     range.
   - If neither binary nor hex, parse normally as a `[BigInteger]`.
2. Apply the multiplier value before attempting any casts to ensure type bounds
   can be appropriately checked without overflows.
3. Check type suffixes.
   - Check type bounds and attempt to parse into that type.
   - If no suffix is used, then the value is bounds-checked in the following
     order, resulting in the first successful test determining the type of the
     number.
     - `[int]`
     - `[long]`
     - `[decimal]` (base-10 literals only)
     - `[double]` (base-10 literals only)
   - If the value is outside the `[long]` range for hex and binary numbers, the
     parse fails.
   - If the value is outside the `[double]` range for base 10 number, the parse
     fails.

>    - Higher values must be explicitly written using the `n` bigint-suffix.

Reviewer: ```suggestion
   - Higher values must be explicitly written using the `n` suffix to parse the literal as a `BigInteger`.
```

Author: fixed
